pkg/controllers/provisioning/v1alpha1/reallocator/filter.go

```diff
@@ -25,40 +28,75 @@ type Filter struct {
 	kubeClient client.Client
 }
 
-// Gets Nodes fitting some underutilization
-// TODO: add predicate
-func (f *Filter) GetUnderutilizedNodes(ctx context.Context) ([]*v1.Node, error) {
-	nodeList := &v1.NodeList{}
+// Gets Nodes fitting some underutilization predicate
+func (f *Filter) GetUnderutilizedNodes(ctx context.Context, provisioner *v1alpha1.Provisioner) ([]*v1.Node, error) {
 	underutilized := []*v1.Node{}
 
-	// 1. Get all nodes
-	if err := f.kubeClient.List(ctx, nodeList); err != nil {
-		return nil, fmt.Errorf("listing nodes, %w", err)
+	// 1. Get all provisioner labeled nodes
+	nodeList, err := f.getNodes(ctx, provisioner)
+	if err != nil {
+		return nil, err
 	}
 
 	// 2. Get nodes and the pods on each node
 	for _, node := range nodeList.Items {
 		pods, err := f.getPodsOnNode(ctx, node.Name)
 		if err != nil {
-			return []*v1.Node{}, fmt.Errorf("filtering pods on nodes, %w", err)
+			return []*v1.Node{}, err
 		}
-		if f.isUnderutilized(&node, pods) {
+		// Only checks if it has 0 non-daemon pods right now
+		if f.isUnderutilized(pods) {
 			underutilized = append(underutilized, &node)
 		}
 	}
 	return underutilized, nil
 }
 
+func (f *Filter) GetExpiredNodes(ctx context.Context, provisioner *v1alpha1.Provisioner) ([]*v1.Node, error) {
+	nodeList, err := f.getNodes(ctx, provisioner)
+	if err != nil {
+		return nil, err
+	}
+	nodes := []*v1.Node{}
+	for _, node := range nodeList.Items {
+		ok, err := nodeutils.IsPastTTL(&node)
+		if err != nil {
+			return nodes, fmt.Errorf("checking TTL, %w", err)
+		}
+		if ok {
+			nodes = append(nodes, &node)
+		}
+	}
+	return nodes, nil
```

Comment:
I do wonder if this func should have a bit non-standard error handling? Like the case I'm worried about is where one node is causing `nodeutils.IsPastTTL` to error out, so you never get to the remaining nodes (ever) until a human manually fixes it (even though the rest of them would not fail `IsPastTTL`)? Another option would be to randomize the walk thru `range nodeList.Items` so that you'll eventually get to the expired nodes (I'm assuming this whole call chain is retried periodically?)

Reply:
I had a similar conversation with Prateek about the `GetUnderUtilizedNodes()` function. We decided to continue with the function and log that the node errored out.